asreview/webapp/authentication/models.py

```diff
@@ -0,0 +1,272 @@
+# Copyright 2019-2022 The ASReview Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime as dt
+from pathlib import Path
+from uuid import uuid4
+
+from flask_login import UserMixin
+from itsdangerous import URLSafeTimedSerializer
+from sqlalchemy import Boolean
+from sqlalchemy import Column
+from sqlalchemy import DateTime
+from sqlalchemy import ForeignKey
+from sqlalchemy import Integer
+from sqlalchemy import String
+from sqlalchemy.orm import relationship
+from sqlalchemy.orm import validates
+from sqlalchemy.orm.session import Session
+from werkzeug.security import check_password_hash
+from werkzeug.security import generate_password_hash
+
+import asreview.utils as utils
+from asreview.webapp import DB
+
+
+class User(UserMixin, DB.Model):
+    """The User model for user accounts."""
+
+    __tablename__ = "users"
+    id = Column(Integer, primary_key=True)
+    identifier = Column(String(100), nullable=False)
+    origin = Column(String(100), nullable=False)
+    email = Column(String(100), unique=True)
+    name = Column(String(100))
+    affiliation = Column(String(100))
+    hashed_password = Column(String(100), unique=True)
+    confirmed = Column(Boolean)
+    public = Column(Boolean)
+    token = Column(String(50))
+    token_created_at = Column(DateTime)
+
+    projects = relationship("Project", back_populates="owner", cascade="all, delete")
+
+    involved_in = relationship(
+        "Project", secondary="collaborations", back_populates="collaborators"
+    )
+
+    pending_invitations = relationship(
+        "Project",
+        secondary="collaboration_invitations",
+        back_populates="pending_invitations",
+    )
+
+    @validates("identifier")
+    def validate_identifier(self, _key, identifier):
+        if not bool(identifier):
+            raise ValueError("Identifier is required")
+        return identifier
+
+    @validates("origin")
+    def validate_origin(self, _key, origin):
+        if not bool(origin):
+            raise ValueError("Origin is required")
+        return origin
+
+    @validates("name")
+    def validate_origin(self, _key, name):
+        if not bool(name):
+            raise ValueError("Name is required")
+        return name
+
+    # this is pretty weird, it 'loops', starting at
+    # origin, stores origin because of the return, moves
+    # on to email and now self contains origin so
+    # I can compare it... and so on.
```

Comment:
Could you explain what is happening here?

Reply:
Is this also where password requirements are checked?

Reply:
No, this comment should have been removed. I made a note because the thing was curious, and I forgot to remove it. Password requirements are not checked here! I am not sure password requirements are checked in the back-end! But in the table we only store the hashed password which doesn't require a validation other than being there if the origin of the account is 'asreview'.